Follow-ups from #1 review

glasser opened the pull request:
I've implemented most of the changes I suggested last week in #1. Each commit implements a different suggestion.

The only remaining comment not addressed here is https://github.com/apollo-specs/core/pull/1#discussion_r592760529

Note that the draft is visible at https://apollo-specs.github.io/core/draft/glasser-20210318/ based on the name chosen for the branch.

glasser committed:
example: use correct location

glasser committed:
substantive: ban prefixes ending in underscore

glasser committed:
substantive: eliminate references to "GraphQL identifier"

This concept does not exist in the GraphQL spec. Replace with references to
GraphQL names. Make it clear that `@feature__24hours` is not allowed (the
previous text was ambiguous).

glasser committed:
clarity: avoid "will be renamed" wording

I found this confusing because it suggests that an intermediate processor may
actually change the schema in some way.

glasser committed:
substantive: don't need to define all locations

In previous versions of the spec, this was expressed in a more general context
about all directives you might define in your own core feature.

glasser committed:
clarity: avoid use of undefined term "default name"

Also another change to be clear that processors are changing their
interpretation of a spec rather than actually mutating the spec.

glasser committed:
example: fix use of @core in all examples

Ensure all examples have a correct self-application of `@core` and a definition
for the directive.

glasser committed:
example: fix single underscores in an example

glasser committed:
clarity: include `v` in sample version tag

glasser committed:
clarity: capitalize sentence

glasser committed:
substantive: remove Ordering section

In the context of this document, ordering is only used as part of a suggestion
about how authors may choose which version to reference, but I'm not even
convinced that this is a generally applicable suggestion. Instead of defining a
whole subsection and algorithm just for this one suggestion, I changed the
suggestion to be a bit broader.

glasser committed:
substantive: require the bootstrap @core to be first

This matches text earlier in the spec which says "The first {@core} directive on
the schema MUST reference the core spec itself, i.e. this document"

glasser committed:
formatting: put fails/warns in bulleted lists

The old formatting had no line breaks and was being rendered as a single
paragraph.

glasser committed:
clarity: fix formatting on 'coreName'

curlies = italics = name of a variable; backticks = code = literal text

trevor-scheer commented on core.spec.md:
```diff
@@ -278,16 +281,19 @@ If true, the element is always exported, regardless of whether the feature which
+  4. MUST NOT end with an underscore (which would create ambiguity between whether {"x___y"} is prefix `x_` for element `y` or prefix `x` for element `_y`).
 
-Prefixed names consist of the name of the feature, followed by two underscores, followed by the name of the element (which can be any valid GraphQL identifier). For instance, the `core` specification (which you are currently reading) introduces elements named [{@core}](#@core) [{@core__export}](#@core__export).
+Prefixed names consist of the name of the feature, followed by two underscores, followed by the name of the element, which can be any valid [GraphQL name](https://spec.graphql.org/draft/#Name). (Note that both parts must be valid GraphQL names, and GraphQL names cannot start with digits, so core feature specifications cannot introduce names like `@feature__24hours`.) For instance, the `core` spec ...
```
>(Note...)

This whole bit feels either like a dedicated aside or just a regular sentence. It seems odd to start a sentence with such a long aside, or whatever this is trying to be.

glasser replied:
I agree that it's a bit odd. Any thoughts on how it could be better?

trevor-scheer replied:
I think this is perfectly fine as a regular sentence. I think I like it better as an aside, after the block, though:

Prefixed names consist of the name of the feature, followed by two underscores, followed by the name of the element, which can be any valid [GraphQL name](https://spec.graphql.org/draft/#Name). For instance, the `core` specification (which you are currently reading) introduces elements named [{@core}](#@core) [{@core__export}](#@core__export).

> Note that both parts must be valid GraphQL names, and GraphQL names cannot start with digits, so core feature specifications cannot introduce names like `@feature__24hours`.

glasser replied:
I don't see any other examples of "blockquote"/"notes" in the doc so I'll just move it to its own paragraph like you did, but without special formatting.

trevor-scheer commented on core.spec.md:
```diff
@@ -351,24 +358,10 @@ As specified by semver, spec authors SHOULD increment the:
 
 As with [semver](https://semver.org/spec/v2.0.0.html), the `0.x` version series is special: there is no expectation of compatibility between versions `0.x` and `0.y`. For example, a processor must not activate implementation `0.4` to satisfy a requested version of `0.2`.
 
-## Ordering
```
Is this removed because it feels unnecessary or extraneous, or some other reason?
I don't quite follow the point you're expressing [here](https://github.com/apollo-specs/core/pull/1/files#r592765808).

glasser replied:
Yes, it's extraneous in my opinion. I had previously gone into more detail at https://github.com/apollo-specs/core/pull/1#discussion_r571311274

Satisfaction actually affects processing, but this ordering relation only affects the suggestion to choose a low-ordered version, which I don't find super compelling and changed.

trevor-scheer replied:
Yeah that makes sense to me, esp. that satisfaction is really the main concern here.

trevor-scheer commented on core.spec.md:
```diff
@@ -380,7 +373,7 @@ Satisfies(requested, available) :
 ## Referencing versions and activating implementations
 
-Schema documents MUST reference a feature version which supports all the schema elements and behaviors required by the document. As a practical matter, authors should generally prefer to reference the [lowest](#sec-Ordering) versions which can [satisfy](#sec-Satisfaction) their requirements, as this imposes the fewest constraints on processors and is thus more likely to be supported. Authors MAY  ...
+Schema documents MUST reference a feature version which supports all the schema elements and behaviors required by the document. As a practical matter, authors will generally prefer to reference a version they have reason to believe is supported by the most processors; depending on context, this might be an old stable version with a low major version, or a new less-deprecated version with a large ...
```
Everything after the first sentence just feels speculative and doesn't really pertain to the spec, I don't think. We should keep it if it actually seems helpful, but I'm unsure.

glasser replied:
Yeah, this is in line with me thinking that offering specific suggestions based on ordering seemed speculative, I think. Should we just remove the entire section or something? Or leave in one sentence maybe somewhere else?

trevor-scheer replied:
I could see this sentence being transformed a bit to provide example / be more informative than speculative.

>Practically speaking, a spec author might prefer to reference a `@core` version which they have reason to believe is supported by most processors; depending on the requirements of the spec, this might be an old stable version with a low major version, or a new less-deprecated version with a large major version.

In re-reading this, I'm not actually sure it makes sense the way it talks about low vs. large major versions, right? Generally speaking, the way an author writes their spec will sorta dictate the major version that applies I think?

glasser replied:
Ah, maybe this isn't clear: this section is (I think!) about people writing schemas, not people writing specs.  So it's trying to tell you "given a choice between referencing `join` v2.3 and `join` v3.5, you should prefer... um... well honestly it probably depends"

trevor-scheer commented on core.spec.md:
```diff
@@ -431,9 +424,10 @@ Determine the name of the core specification within the document.
-**Warns** *Extra Schema* for extra SchemaDefinitions in the document after the first
-**Fails** the *Has Core Feature* validation if the core feature is not referenced with a {@core} directive within the document.
+- **Fails** the *Has Schema* validation if there are no SchemaDefinitions in the document
+- **Warns** *Extra Schema* for extra SchemaDefinitions in the document after the first
```
Warning why? This seems like a fail scenario. What's expected behavior when 2 schema defs are provided? Which do we use?

trevor-scheer replied:
Ok I see the behavior is defined in the bootstrap algorithm below. I'm not sure if I like it though, it just seems like a simple and reasonable thing to fail on.

glasser replied:
I'm just doing reformatting here. Earlier in the process I asked about error handling and Ashi added this fail/warn concept.

Actually, this one is invalid GraphQL (spec says "When using the type system definition language, a document must include at most one schema definition.") so it seems like this should be a strong failure. That leaves only one warning, which is for invalid feature URL. Shall we just make that a failure too and remove the concept of warnings? Certainly simpler.

trevor-scheer replied:
Yeah. Hm, I could've sworn I found the same thing you just did last time I looked - but didn't find it when I looked today. In that case I'd definitely opt for all fails and removing this validation since it's extraneous. This has some cascading changes to go along with it (the algorithm below is the only thing that comes right to mind).

trevor-scheer commented on core.spec.md:
```diff
@@ -431,9 +424,10 @@ Determine the name of the core specification within the document.
+- **Fails** the *Has Schema* validation if there are no SchemaDefinitions in the document
+- **Warns** *Extra Schema* for extra SchemaDefinitions in the document after the first
+- **Fails** the *Has Core Feature* validation if the `core` feature itself is not referenced with a {@core} directive within the document
+- **Fails** the *Bootstrap Core Feature Listed First* validation if the reference to the `core` feature is not the first {@core} directive on the document's SchemaDefinition
```
This rule is important because we need to be able to identify if `core` was renamed?

glasser replied:
My main goal was to make the algorithm consistent with the text of the spec.

I do think it's reasonable though to say "if you're using core, which requires you to declare that you're using core, you should do so first". (Note that this means "the first `@core` has to be the `core` one" but not "the first of all directives on the schema has to be this one".)

trevor-scheer replied:
Yep this is precisely how I interpreted it (including the last bit re: non-core directives)

trevor-scheer commented on core.spec.md:
```diff
@@ -443,27 +437,27 @@ Bootstrap(document) :
     - &#8230;{d} has an [`as:`](#@core/as) argument whose value is equal to {d}'s name
     - &#8230;*or* {d} does not have an [`as:`](#@core/as) argument and {d}'s name is `core`
-    - *then* **Return** {d}'s name
+    - *then* if any directive on {schema} listed before {d} has the same name as {d}, the ***Bootstrap Core Feature Listed First* validation fails**; otherwise **Return** {d}'s name
```
This took me a long time to grok and finally decide it's correct, but partly compounded by groking the previous steps as well.

glasser replied:
Yeah, I think the layout here didn't help the situation. Do you have thoughts on a way of presenting this that would be easier to digest?

trevor-scheer replied:
Having a hard time with thinking up something easier to digest. I'm ok to leave it as is - it does make sense.

Wonder if someone else might have a suggestion. Also, do we have any interest in getting a review from Stephen? Doesn't even have to be blocking, but a follow-up from him would probably yield some great refinements.

trevor-scheer commented on core.spec.md:
```diff
@@ -443,27 +437,27 @@ Bootstrap(document) :
-    - Let {name} be the spec's [name](#sec-Prefixing) as specified by the directive's [`as:`](#@core/as) argument or, if the argument is not present, {defaultName}
-    - If {name} exists within {features}, the ***Name Uniqueness* validation fails**.
-    - Insert {name} => {d} into {features}
+    - Let {featureName} be the {d}'s [`as:`](#@core/as) argument or, if the argument is not present, {specifiedFeatureName}
+    - If {featureName} exists within {features}, the ***Name Uniqueness* validation fails**.
+    - Insert {featureName} => {d} into {features}
```
This is a nice clarifying addition to the naming

glasser committed:
move sentence added in this PR to its own paragraph

glasser committed:
substantive: remove concept of warnings

There are only two warnings. One involves ignoring a case that is invalid
GraphQL so it shouldn't be ignored. The other is more arguably useful as a
warning, but for simplicity let's just do strict error handling consistently; we
can add warnings back in a future spec version.

trevor-scheer commented on core.spec.md:
```diff
@@ -418,23 +418,21 @@ This section lays out algorithms for processing core schemas.
 
 - **Fails** the *Has Schema* validation if there are no SchemaDefinitions in the document
-- **Warns** *Extra Schema* for extra SchemaDefinitions in the document after the first
+- **Fails** *Extra Schema* for extra SchemaDefinitions in the document after the first (note that is is also [invalid GraphQL](http://spec.graphql.org/draft/#sec-Root-Operation-Types)
```
Can't we literally get rid of this since it's not a concern of `core` but of valid graphql itself? Like we don't also have to say things like "you can't have duplicate type names" and every other validation that the graphql spec enforces.

trevor-scheer commented on core.spec.md:
```diff
@@ -418,23 +418,21 @@ This section lays out algorithms for processing core schemas.
 1. Let {schema} be the only SchemaDefinition in {document}
   1. ...if no SchemaDefinitions are present in {document}, the ***Has Schema* validation fails**.
-  2. ...if multiple SchemaDefinitions are present in {document}, let {schema} be the first SchemaDefinition in {document}. Issue the ***Extra Schema* warning** for each subsequent SchemaDefinition, and ignore them for all further processing.
+  2. ...if multiple SchemaDefinitions are present in {document}, the ***Extra Schema* validation fails**.
```
If we do get rid of the multiple schema def validation then we can get rid of this line too.

glasser committed:
no formal validation failure for illegal GraphQL

It is helpful for implementors to be reminded that they don't have to worry
about multiple `schema` definitions, but we don't need to duplicate this
GraphQL-level validation as a Core Schema validation.
